lit_gpt/model_transformer_engine.py

```diff
@@ -0,0 +1,257 @@
+from typing import List, Optional, Tuple
+
+import torch
+import torch.nn as nn
+from torch.nn import functional as F
+from transformer_engine.pytorch import *
+
+from lit_gpt.config import Config
+from lit_gpt.model import apply_rope, GPT as BaseModel, RoPECache, KVCache
+from lit_gpt.utils import find_multiple
+
+
+ATTN_MASK_TYPE = "causal"  # use "padding" for inference
+
+
+class GPT(BaseModel):
+    def __init__(self, config: Config) -> None:
+        nn.Module.__init__(self)
+        assert config.padded_vocab_size is not None
+        self.config = config
+
+        self.wte = nn.Embedding(config.padded_vocab_size, config.n_embd)
+        self.transformer = nn.ModuleDict({"h": nn.ModuleList(Block(config, i) for i in range(config.n_layer))})
+        self.ln_f_lm_head = LayerNormLinear(
+            config.n_embd, config.padded_vocab_size, bias=False, params_dtype=torch.get_default_dtype()
+        )
+
+        self.rope_cache: Optional[RoPECache] = None
+        self.mask_cache: Optional[torch.Tensor] = None
+        self.kv_caches: List[KVCache] = []
+
+    def _init_weights(self, module: nn.Module) -> None:
+        if isinstance(module, (Linear, LayerNormLinear)):
+            # https://huggingface.co/stabilityai/stablelm-base-alpha-3b/blob/main/config.json#L10
+            torch.nn.init.normal_(module.weight, mean=0.0, std=0.02)
+            if module.bias is not None:
+                torch.nn.init.zeros_(module.bias)
+        elif isinstance(module, nn.Embedding):
+            torch.nn.init.normal_(module.weight, mean=0.0, std=0.02)
+        elif isinstance(module, LayerNorm):
+            torch.nn.init.ones_(module.weight)
+            torch.nn.init.zeros_(module.bias)
+            # https://huggingface.co/stabilityai/stablelm-base-alpha-3b/blob/main/config.json#L12
+            module.eps = 1e-5
+        if isinstance(module, (LayerNormLinear, LayerNormMLP)):
+            torch.nn.init.ones_(module.layer_norm_weight)
+            torch.nn.init.zeros_(module.layer_norm_bias)
+        if isinstance(module, LayerNormMLP):
+            torch.nn.init.normal_(module.fc1_weight, mean=0.0, std=0.02)
+            if module.fc1_bias is not None:
+                torch.nn.init.zeros_(module.fc1_bias)
+            torch.nn.init.normal_(module.fc2_weight, mean=0.0, std=0.02)
+            if module.fc2_bias is not None:
+                torch.nn.init.zeros_(module.fc2_bias)
+
+    def forward(
+        self,
+        idx: torch.Tensor,
+        max_seq_length: Optional[int] = None,
+        input_pos: Optional[torch.Tensor] = None,
+        padding_multiple: int = 8,
+    ) -> torch.Tensor:
+        B, T = idx.size()
+        use_kv_cache = input_pos is not None
+
+        T_padded = find_multiple(T, padding_multiple)
```

Comment:
This assumes that all sequences in the batch are good up to T, which might not be true for batched inference